### ISSUE TITLE
Problem with open ESRI Shapefile using gdal.OpenEx from in-memory zip

### ISSUE THREAD

**Issue by artrey** [OSGeo/gdal#8031]
## Expected behavior and actual behavior.

I ran this code on different `osgeo/gdal:ubuntu-small-3.x.x` images. And not all of them execute the code correctly:

- 3.2.3: works
- 3.3.2: works
- 3.4.1: doesn't work
- 3.4.3: doesn't work
- 3.5.3: doesn't work
- 3.6.3: doesn't work
- 3.7.0: doesn't work

Looks like since version 3.4 there are some problems...

```python
import json
import os
import zipfile
from io import BytesIO

from osgeo import gdal


def read_file(file, mode="r"):
    vsi_file = gdal.VSIFOpenL(file, mode)

    if vsi_file is None:
        raise FileNotFoundError(2, "No such file or directory", file)

    gdal.VSIFSeekL(vsi_file, 0, 2)
    vsi_len = gdal.VSIFTellL(vsi_file)
    gdal.VSIFSeekL(vsi_file, 0, 0)

    content = gdal.VSIFReadL(1, vsi_len, vsi_file)
    gdal.VSIFCloseL(vsi_file)

    return content


geojson = {
    "type": "FeatureCollection",
    "features": [
        {
            "type": "Feature",
            "properties": {"prop1": 10, "prop2": 20},
            "geometry": {
                "coordinates": [
                    [
                        [13.369741347157202, 52.591904276169515],
                        [13.369741347157202, 52.58253073291863],
                        [13.384851590246882, 52.58253073291863],
                        [13.384851590246882, 52.591904276169515],
                        [13.369741347157202, 52.591904276169515],
                    ]
                ],
                "type": "Polygon",
            },
        }
    ],
}

vsi_mem = "/vsimem/input.geojson"
gdal.FileFromMemBuffer(vsi_mem, json.dumps(geojson).encode("utf-8"))
input_ds = gdal.OpenEx(vsi_mem, gdal.GA_ReadOnly)
if not input_ds:
    raise RuntimeError("input error")

output_format = "shp"
output_name = "output"
file_path = f"/vsimem/{output_name}"
vsi_mem = f"{file_path}.{output_format}"

gdal.VectorTranslate(vsi_mem, input_ds, format="ESRI Shapefile")

files = []
extensions = {".shp", ".dbf", ".shx"}
for extension in extensions:
    filename = f"{file_path}{extension}"
    try:
        file = read_file(filename)
        files.append((file, filename))
    except FileNotFoundError:
        pass

with BytesIO() as stream:
    path = os.path.dirname(files[0][1])

    with zipfile.ZipFile(stream, mode="w", compression=zipfile.ZIP_DEFLATED) as zip:
        for uploaded_file, uploaded_filename in files:
            zip.writestr(uploaded_filename, uploaded_file)

    vsi_mem = "/vsimem/data.zip"
    vsi = "/vsizip/{0}/{1}".format(vsi_mem, path)
    gdal.FileFromMemBuffer(vsi_mem, stream.getvalue())

output_ds = gdal.OpenEx(vsi, gdal.GA_ReadOnly)
if not output_ds:
    raise RuntimeError("output error")

print("OK:", output_ds.GetLayer(0).GetLayerDefn().GetFieldCount())
```

## Steps to reproduce the problem.

Just run code using python (tested on python3.8+)

```bash
docker run --rm -v test_gdal.py:/src/demo.py osgeo/gdal:ubuntu-small-3.4.3 python /src/demo.py

Traceback (most recent call last):
  File "/src/demo.py", line 82, in <module>
    raise RuntimeError("output error")
RuntimeError: output error
```

```bash
docker run --rm -v test_gdal.py:/src/demo.py osgeo/gdal:ubuntu-small-3.3.2 python /src/demo.py 

OK: 2
```

## Operating system

Ubuntu 20.04 / Ubuntu 22.04 (`osgeo/gdal:ubuntu-small-3.x.x` docker images )

## GDAL version and provenance

- 3.2.3: works
- 3.3.2: works
- 3.4.1: doesn't work
- 3.4.3: doesn't work
- 3.5.3: doesn't work
- 3.6.3: doesn't work
- 3.7.0: doesn't work

**Comment by artrey** [comment]
I figured out what the problem is and how to solve it :)

---

#### TL;DR

A leading slash in a zip-archive is evil :imp:

---

I have a zip-archive with a leading slash:

```bash
>>> zipinfo data2.zip

Archive:  data2.zip
Zip file size: 540 bytes, number of entries: 3
?rw-------  2.0 unx      117 b- defN 23-Jun-30 19:04 /vsimem/output.dbf
?rw-------  2.0 unx      236 b- defN 23-Jun-30 19:04 /vsimem/output.shp
?rw-------  2.0 unx      108 b- defN 23-Jun-30 19:04 /vsimem/output.shx
3 files, 461 bytes uncompressed, 182 bytes compressed:  60.5%
```

And I have to write this leading slash while reading:

```python
from osgeo import gdal

ds = gdal.OpenEx("/vsizip/data2.zip//vsimem", gdal.GA_ReadOnly)
```

But I get an empty `ds = None` for the gdal 3.4+ and everything is OK for the gdal 3.3-  :scream:

After it I repack the archive without the leading slash:

```bash
>>> zipinfo data.zip

Archive:  data.zip
Zip file size: 534 bytes, number of entries: 3
?rw-------  2.0 unx      236 b- defN 23-Jun-30 20:03 vsimem/output.shp
?rw-------  2.0 unx      108 b- defN 23-Jun-30 20:03 vsimem/output.shx
?rw-------  2.0 unx      117 b- defN 23-Jun-30 20:03 vsimem/output.dbf
3 files, 461 bytes uncompressed, 182 bytes compressed:  60.5%
```

Now I have to write only the folder name without the leading slash:

```python
from osgeo import gdal

ds = gdal.OpenEx("/vsizip/data.zip/vsimem", gdal.GA_ReadOnly)
```

And this code works fine using gdal 3.0+ :partying_face:

**Comment by jratike80** [comment]
I consider that this really was not a bug at least if interoperability is needed. I tried to create a zip with such filenames on Windows but I did not manage to do it because slash is a reserved character in filenames. But if you have such a zip file it would be interesting to test if Windows can find the files at all. Can you test that on Windows, or share a dummy test zip as an annex to this issue?

**Comment by artrey** [comment]
Here is zip with the leading slash (it's temporary storage, so the file will be available only 30 days):

https://temp-file.org/KWlUSPXjgS0FfB3/file

**Comment by jratike80** [comment]
The "Extract All" function of the standard Windows 10 File Explorer gives an error and says that the zip archive is invalid. It seems that using leading slashes is not a good idea.

However, 7-zip on Windows could unzip the file and it placed the shapefile into a subdirectory "vsimem".

Would you mind editing the title of this issue so that the leading slash is mentioned? Other users may meet the same problem and then they could find the issue and comments easier.
